File: src/api/searchQuality.ts
// Mirrors backend/search_quality_routes.py, backend/search_query.py, and backend/search_quality.py.
import { normalizeDatabaseName } from './search';

export type SyntaxInfo = { key: string; label: string; databases: string };

export type PressElement = { key: string; label: string; guidance: string };

export type PressRating = 'no_revision' | 'revision_suggested' | 'revision_required';

export type SearchQualityCatalog = {
  syntaxes: SyntaxInfo[];
  press: { elements: PressElement[]; ratings: PressRating[] };
  recall_connectors: string[];
  database_syntaxes: Record<string, string>;
};

export type ValidationResult = { syntax: string; valid: boolean; errors: string[]; warnings: string[]; term_count: number | null };

export type Translation = { target: string; target_label: string; query: string; warnings: string[] };

export type MeshHeading = { ui: string; heading: string; entry_terms: string[]; scope_note: string; tree_numbers: string[] };

export type RecallCheck = {
  id: number;
  connector: string;
  strategy_version: number;
  seeds: string[];
  found: string[];
  missed: string[];
  recall: number | null;
  created_by: string | null;
  created_at: string;
};

export type PressReview = {
  id: number;
  strategy_id: number | null;
  strategy_version: number | null;
  status: 'completed' | 'waived';
  answers: Record<string, { rating: PressRating; comment: string }>;
  overall: 'approved' | 'revisions_required' | null;
  comment: string;
  reviewer: string | null;
  created_at: string;
};

export type PressStrategyStatus = {
  strategy_id: number;
  database: string;
  version: number;
  status: 'approved' | 'revisions_required' | 'outdated' | 'not_reviewed';
  review_id: number | null;
};

export type PressStatus = { waived: boolean; waiver_reason: string | null; strategies: PressStrategyStatus[]; met: boolean };

export type StrategyVersion = { version: number; database: string; query: string; note: string | null; created_by: string | null; created_at: string };

export const PRESS_RATING_LABELS: Record<PressRating, string> = {
  no_revision: 'No revision',
  revision_suggested: 'Revision suggested',
  revision_required: 'Revision required',
};

export const PRESS_STATUS_LABELS: Record<PressStrategyStatus['status'], string> = {
  approved: 'PRESS approved',
  revisions_required: 'PRESS: revisions required',
  outdated: 'PRESS review is for an older version',
  not_reviewed: 'Not peer reviewed',
};

export const syntaxForDatabase = (catalog: SearchQualityCatalog | null, database: string) =>
  catalog?.database_syntaxes[normalizeDatabaseName(database)] ?? null;

File: backend/search_quality.py
"""Search quality: strategy versions, PRESS 2015 peer review, and recall checks against known relevant articles."""

from dataclasses import asdict, dataclass

from sqlalchemy import select
from sqlalchemy.orm import Session

import models
from services.literature_sources import europepmc_found_seeds
from services.openalex import openalex_found_seeds
from services.pubmed import pubmed_found_seeds
from services.record_import import normalize_doi


@dataclass(frozen=True)
class PressElement:
    key: str
    label: str
    guidance: str


# The six elements of the PRESS 2015 Evidence-Based Checklist (McGowan et al., J Clin Epidemiol 2016;75:40-46).
PRESS_ELEMENTS = (
    PressElement(
        "translation",
        "Translation of the research question",
        "Does the search match the research question? Are there too many or too few concepts, or wrong combinations?",
    ),
    PressElement(
        "operators",
        "Boolean and proximity operators",
        "Are AND, OR, and NOT used correctly, and could NOT remove relevant records? Is proximity used where it helps?",
    ),
    PressElement(
        "subject_headings",
        "Subject headings",
        "Are headings relevant and complete, correctly exploded or not, and paired with free-text terms?",
    ),
    PressElement(
        "text_words",
        "Text word searching",
        "Are synonyms, spelling variants, abbreviations, and truncation covered, in the right fields?",
    ),
    PressElement(
        "spelling_syntax",
        "Spelling, syntax, and line numbers",
        "Are there spelling mistakes, syntax errors for the interface, or wrong line combinations?",
    ),
    PressElement(
        "limits_filters",
        "Limits and filters",
        "Are limits and filters (dates, languages, designs, humans) justified by the protocol and applied correctly?",
    ),
)
PRESS_KEYS = [element.key for element in PRESS_ELEMENTS]
PRESS_RATINGS = ("no_revision", "revision_suggested", "revision_required")
RECALL_CONNECTORS = ("pubmed", "europepmc", "openalex")


def press_catalog() -> dict:
    return {"elements": [asdict(element) for element in PRESS_ELEMENTS], "ratings": list(PRESS_RATINGS)}


def record_strategy_version(
    db: Session, strategy: models.SearchStrategy, user_id: int | None, note: str | None
) -> None:
    db.add(
        models.SearchStrategyVersion(
            strategy_id=strategy.id,
            version=strategy.version,
            database=strategy.database,
            query=strategy.query,
            note=note,
            created_by_id=user_id,
        )
    )


def version_author(db: Session, strategy: models.SearchStrategy) -> int | None:
    return db.scalar(
        select(models.SearchStrategyVersion.created_by_id).where(
            models.SearchStrategyVersion.strategy_id == strategy.id,
            models.SearchStrategyVersion.version == strategy.version,
        )
    )


def press_status(db: Session, project_id: int) -> dict:
    """PRESS status of each strategy's current version, and whether the search gate's PRESS requirement is met."""
    reviews = db.scalars(
        select(models.PressReview).where(models.PressReview.project_id == project_id).order_by(models.PressReview.id)
    ).all()
    waiver = next((review for review in reversed(reviews) if review.status == "waived"), None)
    strategies = db.scalars(
        select(models.SearchStrategy)
        .where(models.SearchStrategy.project_id == project_id)
        .order_by(models.SearchStrategy.id)
    ).all()
    items = []
    for strategy in strategies:
        completed = [r for r in reviews if r.strategy_id == strategy.id and r.status == "completed"]
        current = [r for r in completed if r.strategy_version == strategy.version]
        if current:
            status = "approved" if current[-1].overall == "approved" else "revisions_required"
        else:
            status = "outdated" if completed else "not_reviewed"
        items.append(
            {
                "strategy_id": strategy.id,
                "database": strategy.database,
                "version": strategy.version,
                "status": status,
                "review_id": current[-1].id if current else None,
            }
        )
    return {
        "waived": waiver is not None,
        "waiver_reason": waiver.comment if waiver else None,
        "strategies": items,
        "met": waiver is not None or (bool(items) and all(item["status"] == "approved" for item in items)),
    }


def parse_seeds(seeds: list[str]) -> tuple[list[str], list[str], list[str]]:
    """Split known relevant articles into PMIDs and DOIs; anything else is returned as unrecognized."""
    pmids, dois, unrecognized = [], [], []
    for seed in seeds:
        value = seed.strip()
        if not value:
            continue
        if value.isdigit():
            pmids.append(value)
        elif (doi := normalize_doi(value).lower()).startswith("10."):
            dois.append(doi)
        else:
            unrecognized.append(value)
    return list(dict.fromkeys(pmids)), list(dict.fromkeys(dois)), unrecognized


def found_seeds(connector: str, query: str, pmids: list[str], dois: list[str]) -> set[str]:
    if connector == "pubmed":
        return pubmed_found_seeds(query, pmids, dois)
    if connector == "europepmc":
        return europepmc_found_seeds(query, pmids, dois)
    if connector == "openalex":
        return openalex_found_seeds(query, pmids, dois)
    raise ValueError(f"Recall checks aren't available for {connector}")
